src/components/virtual-gamepad-controls.js

```diff
@@ -0,0 +1,188 @@
+import nipplejs from "nipplejs";
+
+var THREE = AFRAME.THREE;
+var DEGREES = Math.PI / 180;
+var HALF_PI = Math.PI / 2;
+
+AFRAME.registerComponent("virtual-gamepad-controls", {
+  schema: {
+    movementSpeed: { default: 2 },
+    lookSpeed: { default: 60 }
+  },
+
+  init() {
+    // Setup gamepad elements
+    var leftTouchZone = document.createElement("div");
+    leftTouchZone.style.position = "absolute";
+    leftTouchZone.style.left = 0;
+    leftTouchZone.style.top = 0;
+    leftTouchZone.style.bottom = 0;
+    leftTouchZone.style.right = "50%";
+
+    var rightTouchZone = document.createElement("div");
+    rightTouchZone.style.position = "absolute";
+    rightTouchZone.style.left = "50%";
+    rightTouchZone.style.top = 0;
+    rightTouchZone.style.bottom = 0;
+    rightTouchZone.style.right = 0;
+
+    document.body.appendChild(leftTouchZone);
+    document.body.appendChild(rightTouchZone);
+
+    var leftStick = nipplejs.create({
+      zone: leftTouchZone,
+      mode: "static",
+      color: "white",
+      position: { left: "50px", bottom: "50px" }
+    });
+
+    leftStick[0].el.style.margin = "5vh 5vw";
+
+    var rightStick = nipplejs.create({
+      zone: rightTouchZone,
+      mode: "static",
+      color: "white",
+      position: { right: "50px", bottom: "50px" }
+    });
+
+    rightStick[0].el.style.margin = "5vh 5vw";
+
+    this.onJoystickChanged = this.onJoystickChanged.bind(this);
+
+    rightStick.on("move end", this.onJoystickChanged);
+    leftStick.on("move end", this.onJoystickChanged);
+
+    this.leftTouchZone = leftTouchZone;
+    this.rightTouchZone = rightTouchZone;
+    this.leftStick = leftStick;
+    this.rightStick = rightStick;
+
+    // Define initial state
+    this.velocity = new THREE.Vector3();
+    // this.pitch = 0;
+    this.yaw = 0;
+
+    // Allocate matrices and vectors
+    this.move = new THREE.Matrix4();
+    this.trans = new THREE.Matrix4();
+    this.transInv = new THREE.Matrix4();
+    this.pivotPos = new THREE.Vector3();
+    this.rotationAxis = new THREE.Vector3(0, 1, 0);
+    this.yawMatrix = new THREE.Matrix4();
+    this.rotationMatrix = new THREE.Matrix4();
+    this.rotationInvMatrix = new THREE.Matrix4();
+    this.camRotationMatrix = new THREE.Matrix4();
+    this.camRotationInvMatrix = new THREE.Matrix4();
+
+    this.cameraEl = document.querySelector("[camera]");
+
+    this.onEnterVr = this.onEnterVr.bind(this);
+    this.onExitVr = this.onExitVr.bind(this);
+    this.el.sceneEl.addEventListener("enter-vr", this.onEnterVr);
+    this.el.sceneEl.addEventListener("exit-vr", this.onExitVr);
+  },
+
+  onJoystickChanged(event, data) {
```

Comment:
Id use someting other than `data` here as I at first confused it with `this.data`, maybe `details`?